packages/protocol/contracts/bridge/Bridge.sol

```diff
@@ -707,16 +565,57 @@ contract Bridge is EssentialContract, IBridge {
         }
     }
 
-    function _isPostInvocationDelay(
-        uint256 _receivedAt,
-        uint256 _invocationDelay
+    function _calcFee(
+        uint256 _msgFee,
+        uint256 _msgGasLimit,
+        uint256 _gasUsed
     )
         private
         view
-        returns (bool)
+        returns (uint256)
+    {
+        if (_msgFee == 0 || _msgGasLimit == 0) return 0;
+
+        uint256 maxFee = _msgFee * _gasUsed / _msgGasLimit;
+        uint256 baseFee = block.basefee * _gasUsed;
+        return _msgFee.min(baseFee >= maxFee ? maxFee : (maxFee + baseFee) >> 1);
+    }
+
+    function _calcGasToCharge(
+        uint256 _gasStart,
+        uint256 _gasOverhead,
+        uint256 _calldataBytes,
+        uint256 _numCacheOps
+    )
+        private
+        view
+        returns (uint32)
+    {
+        unchecked {
+            uint256 gasUsed = _gasStart - gasleft() + _gasOverhead
+                + _calldataBytes.min(_MAX_PROOF_BYTES_TO_CHARGE) >> 4;
```

Comment:
Agreed. I think we should make sure the GAS_OVERHEAD is large (but reasonable) enough so it will cover most proof data cost. I change the event GasChargedByRelayer to "log" more data so we can calculate GAS_RESERVE and GAS_OVERHEAD more accurately.